catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java

```diff
@@ -0,0 +1,224 @@
+              .includeEmpty()
+              .includeNulls()
+              .includeDefaultValues()
+              .setJsonFormatForDates(false)
+              .useAnnotations());
+
+  private EndpointUtil util;
+
+  public CqlTransformHandler(
+      List<ServiceReference> queryResponseTransformers,
+      BundleContext bundleContext,
+      EndpointUtil endpointUtil) {
+    this.queryResponseTransformers = queryResponseTransformers;
+    this.bundleContext = bundleContext;
+    this.util = endpointUtil;
+  }
+
+  public class Arguments {
+    private Map<String, Object> arguments;
+
+    public Arguments() {
+      this.arguments = Collections.emptyMap();
+    }
+
+    public void setArguments(Map<String, Object> arguments) {
+      this.arguments = arguments;
+    }
+
+    public Map<String, Object> getArguments() {
+      return this.arguments;
+    }
+
+    public Map<String, Serializable> getSerializableArguments() {
+      Map<String, Serializable> serializableMap =
+          this.getArguments()
+              .entrySet()
+              .stream()
+              .filter(entry -> entry.getValue() instanceof Serializable)
+              .collect(Collectors.toMap(Map.Entry::getKey, e -> (Serializable) e.getValue()));
+      return serializableMap;
+    }
+  }
+
+  @Override
+  public Object handle(Request request, Response response) throws Exception {
+    String transformerId = request.params(":transformerId");
+    String body = util.safeGetBody(request);
+    CqlRequest cqlRequest;
+
+    try {
+      cqlRequest = mapper.readValue(body, CqlRequest.class);
+      if (cqlRequest == null) {
+        LOGGER.debug("Cql request parsed from body evaluated to  null.");
+        throw new NullPointerException("Cql request is null.");
+      }
+    } catch (Exception e) {
+      LOGGER.debug("Error fetching cql request");
+      response.status(HttpStatus.BAD_REQUEST_400);
+      return ImmutableMap.of("message", "Bad request");
+    }
+
+    Map<String, Serializable> arguments;
+
+    try {
+      arguments = mapper.readValue(body, Arguments.class).getSerializableArguments();
+    } catch (NullPointerException e) {
+      arguments = Collections.emptyMap();
+    }
+
+    LOGGER.trace("Finding transformer to transform query response.");
+
+    ServiceReference<QueryResponseTransformer> queryResponseTransformer =
+        queryResponseTransformers
+            .stream()
+            .filter(transformer -> transformerId.equals(transformer.getProperty("id")))
+            .findFirst()
+            .orElse(null);
+
+    if (queryResponseTransformer == null) {
+      LOGGER.debug("Could not find transformer with id: {}", transformerId);
+      response.status(HttpStatus.NOT_FOUND_404);
+      return ImmutableMap.of("message", "Service not found");
+    }
+
+    CqlQueryResponse cqlQueryResponse = util.executeCqlQuery(cqlRequest);
+
+    attachFileToResponse(request, response, queryResponseTransformer, cqlQueryResponse, arguments);
+
+    return "";
```

Comment:
❓ This seems strange...the method has an `Object` return type and here we're returning and empty string. Why?

Reply:
@leo-sakh Whatever object is returned is mapped to json. So if there are any failures then a map is returned which will return to the ui as json. 
But in the case of success the response body is overridden with the file content. So even if "success" is sent back it will be overridden and the user won't see it.
The same is used for the CsvTransform endpoint https://github.com/codice/ddf/blob/18e0cb967af4ff722900905a20f6649629e2a0af/catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/metacard/MetacardApplication.java#L633